### PR TITLE
🎉 New source: Open Exchange Rates 

### PR DESCRIPTION
**Pull request by mrhallak**
## What
This is to add a new connector that connects to the [Open Exchanges Rates](https://openexchangerates.org/)

## How
Initialized using the Airbyte CDK.

## Recommended reading order
1. `x.python`

## 🚨 User Impact 🚨
No breaking changes, users should be able to see a new source connector.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>
Screenshot 2022-11-15 at 17 17 03

</details>

<details><summary><strong>Integration</strong></summary>

No integration tests
</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by mrhallak** [comment]
Looking for guidance on (testing with Snowflake as a destination):
- Normal incremental mode produces duplicate rows with the same base and timestamp
- Normal incremental mode produces two records for the same day, but different timestamp, I was wondering if there is a way to transform the timestamp from Unix Epoch to a date and eliminate duplicates if they are on the same day
- Incremental Deduped + history seems to work best but still could result in duplicates because of the unix epoch timestamp

**Review by vincentkoc — DISMISSED** [review]
@mrhallak thanks for your submission, I'm one of the Airbyte maintainers and I’ll take a look at your PR.

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by cptjacky** [comment]
Hi @mrhallak, thank you for your contribution and sorry for the delay!

I will be reviewing your PR soon!

(cc @sajarin)

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates_stream.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates_stream.json:1]: This file is not correctly named: it should match the class name, in your case that would give: `open_exchange_rates.json`

(see [the doc about schema files](https://docs.airbyte.com/connector-development/tutorials/cdk-tutorial-python-http/declare-schema ))

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/sample_config.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/sample_config.json:4]: In your spec.yaml, you defined this key as `app_id`

```suggestion
  "app_id": "Your API Access Key"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/invalid_config.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/invalid_config.json:4]: In your spec.yaml, you defined this key as `app_id`

```suggestion
  "app_id": "wrong-access-key"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:49]: You should use the `TokenAuthenticator` instead of defining it yourself.
(It accepts an `auth_method` parameter that you can set to have `Token` instead of `Bearer`)

See [the doc about auth](https://docs.airbyte.com/connector-development/cdk-python/http-streams#authentication)

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:78]: minor naming nitpick, but this is not really a timestamp but rather a date, it would probably be cleared to rename it to  `date`

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:23]: nitpick: we could just have one property here

```suggestion
    cursor_field = "timestamp"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:53]: In order to properly make this stream `incremental`, you have to implement the state update logic somewhere. The most logical place would likely be here, just before yielding.

The [documentation example](In order to properly make this stream `incremental`, you have to implement the state update logic somewhere) explains how to do this.

I think you should store the `timestamp` returned by the API in the state here, and also make sure what the API returned is `>` (and not equal!) to the current state. If not, you should `return` and not yield the record. This will prevent duplicates in the stream!

File: airbyte-integrations/connectors/source-open-exchange-rates/setup.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/setup.py:8]: I think you should explicitely add `pendulum` to the dependencies here

File: airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml:1]: I think this file is missing the line `connector_image: airbyte/source-retently:dev`

File: airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml:18]: When the connector properly implements the `incremental` functionality, you should re-add the `incremental` acceptance tests here.

See [the acceptance tests doc](https://docs.airbyte.com/connector-development/tutorials/build-a-connector-the-hard-way#4-test-the-connector) for more information

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml
Comment (cptjacky) [inline airbyte-config/init/src/main/resources/seed/source_definitions.yaml:1084]: Can we make this `0.1.0` instead of `dev`?

**Review by cptjacky — COMMENTED** [review]
Hello @mrhallak and thank you for your contribution!

I made some comments that need to be addressed before we can merge this.

If something is unclear, do not hesitate to ask!

File: airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml
Comment (mrhallak) [reply airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml:3]: Added

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates_stream.json
Comment (mrhallak) [reply airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates_stream.json]: Done

File: airbyte-integrations/connectors/source-open-exchange-rates/setup.py
Comment (mrhallak) [reply airbyte-integrations/connectors/source-open-exchange-rates/setup.py:8]: Added

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (mrhallak) [reply airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py]: Made some changes, hope I got this right 😅

**Comment by koconder** [comment]
@mrhallak ill leave it to @cptjacky to finalize your review, please bear with us as some of us are still on leave :)

File: airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml
Comment (cptjacky) [reply airbyte-integrations/connectors/source-open-exchange-rates/acceptance-test-config.yml:20]: @mrhallak, could add the `incremental` acceptance tests to this file?

This should do the job:
```yaml
  incremental:
    tests:
      - config_path: "secrets/config.json"
        configured_catalog_path: "integration_tests/configured_catalog.json"
        future_state:
          future_state_path: "integration_tests/abnormal_state.json"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:23]: In our case, the cursor_field is `timestamp`, not date

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:30]: Let's parse the date into a datetime here: 

```suggestion
        self.start_date = pendulum.parse(start_date)
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:49]: This can removed! The auth headers from the authenticator are automatically used by the cdk

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:56]: Since we're using a timestamp as state, we have to convert the start date to a timestamp if we use it as state.

```suggestion
            return {self.cursor_field: self.start_date.timestamp()}
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:60]: No need to parse it here, we'll already have a timestamp

```suggestion
        self._cursor_value = value[self.cursor_field]
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:69]: Same here, no need to parse it, we'll use the timestamp 

```suggestion
        latest_record_timestamp = response_json['timestamp']
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:72]: if the cursor value doesn't exist, we should set it

```suggestion
        if self._cursor_value:
            self._cursor_value = max(self._cursor_value, latest_record_timestamp)
        else:
            self._cursor_value = latest_record_timestamp
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:77]:
```suggestion
        start_date = pendulum.from_timestamp(stream_state[self.cursor_field] if stream_state and self.cursor_field in stream_state else self.start_date)
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:96]:
```suggestion
            dates.append({"date": start_date.to_date_string()})
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:86]:
```suggestion
        return f"historical/{stream_slice['date']}.json"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/schemas/open_exchange_rates.json:12]: It seems like the returned value is an integer, not a string

```suggestion
            "type": "integer"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/configured_catalog.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/configured_catalog.json:13]:
```suggestion
              "type": "integer"
```

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/configured_catalog.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/configured_catalog.json:1052]: I think append would make more sense than overwrite with incremental!

```suggestion
      "destination_sync_mode": "append",
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:70]: I think we should **not yield** the records if they have the same timestamp as the state, as this means we already yielded it:

```suggestion
        if self._cursor_value and latest_record_timestamp <= self._cursor_value:
            return
```

File: airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/source_open_exchange_rates/source.py:66]: Can we make the signature of this match the one on the abstract class?

```suggestion
    def parse_response(
        self,
        response: requests.Response,
        *,
        stream_state: Mapping[str, Any],
        stream_slice: Mapping[str, Any] = None,
        next_page_token: Mapping[str, Any] = None,
    ) -> Iterable[Mapping]:
```

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/abnormal_state.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/abnormal_state.json:3]: We'll have to put a timestamp here as well

```suggestion
    "timestamp": 2052084644
```

File: airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/sample_state.json
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/integration_tests/sample_state.json:3]: We'll have to put a timestamp here as well

```suggestion
    "timestamp": 1673379575
```

**Review by cptjacky — COMMENTED** [review]
Hello @mrhallak, thank you very much for making the changes!

I made comments to fix a few things and make the incremental work! As always, feel free to ping me if something is unclear :)

**Comment by mrhallak** [comment]
> Hello @mrhallak, thank you very much for making the changes!
> 
> I made comments to fix a few things and make the incremental work! As always, feel free to ping me if something is unclear :)



> Hello @mrhallak, thank you very much for making the changes!
> 
> I made comments to fix a few things and make the incremental work! As always, feel free to ping me if something is unclear :)

Hey @cptjacky thanks for all of the suggestions! I had to add a condition in `stream_slices` to be able to run the below command

`python main.py read --config secrets/config.json --catalog integration_tests/configured_catalog.json --state integration_tests/state.json`

Is there anything I should have done differently here?

File: airbyte-integrations/connectors/source-open-exchange-rates/setup.py
Comment (cptjacky) [inline airbyte-integrations/connectors/source-open-exchange-rates/setup.py:15]: This should be in `MAIN_REQUIREMENTS` instead of `TEST_REQUIREMENTS`

File: docs/integrations/sources/openexchangerates.md
Comment (cptjacky) [inline docs/integrations/sources/openexchangerates.md:1]: I think we should rename this file to match the paths we've used in the connector itself

`open-exchange-rates.md`

**Review by cptjacky — COMMENTED** [review]
Great! Everything in the connector itself looks good!

Two very small comments remain if you could take care of them.

Could you also post a screenshot of the acceptance tests correctly running?

**Comment by mrhallak** [comment]
@cptjacky made the changes, thanks! Below is a screenshot of the running acceptance tests

Screenshot 2023-01-12 at 09 53 29

File: airbyte-integrations/connectors/source-open-exchange-rates/setup.py
Comment (cptjacky) [reply airbyte-integrations/connectors/source-open-exchange-rates/setup.py:12]: Sorry if I was unclear, I meant only `pendulum`, not the others `pytest`/`pytest-mock`

**Comment by cptjacky** [comment]
/publish connector=connectors/source-open-exchange-rates run-tests=false

**Comment by cptjacky** [comment]
/publish connector=connectors/source-open-exchange-rates run-tests=false

**Review by cptjacky — APPROVED** [review]
Thank you @mrhallak for your amazing contribution!

I'll ping the Airbyte team to get this merged :)

**Review by vincentkoc — APPROVED** [review]

**Comment by mrhallak** [comment]
Hello any thoughts when this will be merged? 🙌🏻 @cptjacky

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**
